File: DistC++/Include/Template/HashList.cc
/*
 * Copyright (C) 1995
 *
 * Department of Computing Science,
 * The University,
 * Newcastle upon Tyne,
 * UK.
 *
 * $Id: HashList.cc,v 1.1 1997/09/25 15:27:55 nmcl Exp $
 */

#ifndef HASHLIST_CC_
#define HASHLIST_CC_

/*
 * Ctor/dtor
 */

template <class T>
HashList<T>::HashList ( size_t hashBucket )
		      : maxBucket(hashBucket)
{
    buckets = new BasicList<T> *[hashBucket];
    
    for (int i = 0; i < hashBucket; i++)
	buckets[i] = 0;
}

template <class T>
HashList<T>::~HashList ()
{
    empty();
    
    delete [] buckets;
}

/*
 * Public
 */

template <class T>
Boolean HashList<T>::add ( T *th )
{
    if (th)
    {
	unsigned long entry = th->get_uid().hash() % maxBucket;

	if (!buckets[entry])
	    buckets[entry] = new BasicList<T>;

	return buckets[entry]->add(th);
    }
    
    return FALSE;
}

/*
 * Insert is like add execpt that it only adds once. i.e. if the entry
 * already exists it returns FALSE;
 */

template <class T>
Boolean HashList<T>::insert ( T *th )
{
    if (th)
    {
	unsigned long entry = th->get_uid().hash() % maxBucket;

	if (buckets[entry])
	    return buckets[entry]->insert(th);
	else
	{
	    buckets[entry] = new BasicList<T>;
	    return buckets[entry]->add(th);
	}
    }
    
    return FALSE;
}

template <class T>
Boolean HashList<T>::remove ( T *findMe )
{
    if (findMe)
    {
	unsigned long entry = findMe->get_uid().hash() % maxBucket;

	if (buckets[entry])
	    return buckets[entry]->remove(findMe);
    }
    
    return FALSE;
}

template <class T>
T *HashList<T>::remove ( const Uid& findMe )
{
    unsigned long entry = findMe.hash() % maxBucket;
    
    if (buckets[entry])
	return buckets[entry]->remove(findMe);

    return 0;
}

template <class T>
T *HashList<T>::lookFor ( const Uid& findMe )
{
    unsigned long entry = findMe.hash() % maxBucket;
	
    if (buckets[entry])
	return buckets[entry]->lookFor(findMe);

    return 0;
}

template <class T>
T *HashList<T>::orderedPop ()
{
    T *th = 0;
    
    for (int i = 0; (i < maxBucket) && (!th); i++)
    {
	if (buckets[i] && buckets[i]->size() > 0)
	{
	    th = buckets[i]->orderedPop();
	}
    }
    
    return th;
}

template <class T>
HashListI<T>::HashListI ( const HashList<T> &list )
			: theList(list),
			  iter(0),
			  lastBucket(-1)
{
}

template <class T>
HashListI<T>::~HashListI ()
{
    if (iter)
	delete iter;
}

template <class T>
T *HashListI<T>::operator() ()
{
    T* th = 0;
    
    /*
     * Assume that receiving a null from the current iterator means
     * we have come to the end of that list, so move on to next.
     */

    if (iter)
	th = (*iter)();

    if (th)
	return th;
    else
	iter = 0;

    lastBucket++;  // move on to next bucket!
    
    while ((lastBucket < theList.maxBucket) &&
	   ((theList.buckets[lastBucket] == 0) ||
	    (theList.buckets[lastBucket]->size() == 0)))
	lastBucket++;

    if (lastBucket < theList.maxBucket)
    {
	if (iter != 0)
	    delete iter;
	
	iter = new BasicListI<T>(*theList.buckets[lastBucket]);
    }

    if (iter != 0)
	th = (*iter)();
    
    return th;
}

#endif

File: JTSArjuna/JTSArjuna-2.0_Alpha_2_1/ArjunaCommon/classes/com/arjuna/ArjunaCommon/Template/HashList.java
/*
 * Copyright (C) 1998, 1999, 2000,
 *
 * Arjuna Solutions Limited,
 * Newcastle upon Tyne,
 * Tyne and Wear,
 * UK.
 *
 * $Id: HashList.java,v 1.1 2000/02/25 14:02:55 nmcl Exp $
 */

package com.arjuna.ArjunaCommon.Template;

import com.arjuna.ArjunaCommon.Common.Uid;

/*
 * This hash table for objects which define get_uid does not use
 * the Java Hashtable class, since we sometimes want to add multiple
 * different values under the same key.
 */

public class HashList
{

public HashList (int hashBucket)
    {
	maxBucket = hashBucket;
	buckets = new BasicList[hashBucket];
    
	for (int i = 0; i < hashBucket; i++)
	    buckets[i] = null;
    }

public void finalize ()
    {
	empty();

	buckets = null;
    }

public synchronized void empty ()
    {
	for (int i = 0; i < maxBucket; i++)
	    if (buckets[i] != null)
		buckets[i] = null;
    }

public synchronized int size ()
    {
	int rVal = 0;
    
	for (int i = 0; i < maxBucket; i++)
	{
	    if (buckets[i] != null)
	    {
		rVal += buckets[i].size();
	    }
	}
    
	return rVal;
    }
    
public synchronized boolean add (ListElement th)
    {
	if (th != null)
	{
	    int entry = th.get_uid().hashCode() % maxBucket;
	    
	    if (buckets[entry] == null)
		buckets[entry] = new BasicList();

	    return buckets[entry].add(th);
	}
    
	return false;
    }

    /*
     * Insert is like add execpt that it only adds once, i.e., if the entry
     * already exists it returns false.
     */

public synchronized boolean insert (ListElement th)
    {
	if (th != null)
	{
	    int entry = th.get_uid().hashCode() % maxBucket;

	    if (buckets[entry] != null)
		return buckets[entry].insert(th);
	    else
	    {
		buckets[entry] = new BasicList();
		return buckets[entry].add(th);
	    }
	}
    
	return false;
    }

public synchronized boolean remove (ListElement findMe)
    {
	if (findMe != null)
	{
	    int entry = findMe.get_uid().hashCode() % maxBucket;

	    if (buckets[entry] != null)
		return buckets[entry].remove(findMe);
	}
	
	return false;
    }

public synchronized ListElement remove (Uid findMe)
    {
	int entry = findMe.hashCode() % maxBucket;
    
	if (buckets[entry] != null)
	    return buckets[entry].remove(findMe);

	return null;
    }

public synchronized ListElement lookFor (Uid findMe)
    {
	int entry = findMe.hashCode() % maxBucket;
	
	if (buckets[entry] != null)
	    return buckets[entry].lookFor(findMe);

	return null;
    }

public synchronized ListElement orderedPop ()
    {
	ListElement th = null;
    
	for (int i = 0; (i < maxBucket) && (th != null); i++)
	{
	    if ((buckets[i] != null) && (buckets[i].size() > 0))
	    {
		th = buckets[i].orderedPop();
	    }
	}
    
	return th;
    }

protected BasicList[] buckets;
protected int maxBucket;

};
